### PR TITLE
Fix types in typing.py

### PR DESCRIPTION
**Pull request by jjmurre**
Separate commit to fix the typings in types.py (and utils.py).

Changed some optional parameters (id, parent) for table and field to mandatory (non-keyword) parameters,
because the parameters are always needed anyway. This simplified the `mypy pleasing` quite a bit.

### CODE REVIEW

File: src/schematools/types.py [hunk 1]
```diff
@@ -99,38 +112,38 @@ def url_prefix(self) -> str:
...
 
     @property
-    def identifier(self):
+    def identifier(self) -> str:
         """Which fields acts as identifier. (default is Django "pk" field)"""
         return self.get("identifier", "pk")
 
     @property
-    def version(self):
+    def version(self) -> str:
```
Comment (vdboor) [inline src/schematools/types.py:120]:
```suggestion
    def version(self) -> Optional[str]:
```

File: src/schematools/types.py [hunk 2]
```diff
@@ -99,38 +112,38 @@ def url_prefix(self) -> str:
...
     @property
-    def is_default_version(self):
+    def is_default_version(self) -> bool:
         """Is this Default Dataset version.
         Defaults to True, in order to stay backwards compatible."""
         return self.default_version == self.version
 
     @property
-    def auth(self):
+    def auth(self) -> Optional[Union[str, List[str]]]:
```
Comment (vdboor) [inline src/schematools/types.py:136]: laat maar weg, ik heb hier consistent `set[str]` van gemaakt in de auth refactor.

File: src/schematools/types.py [hunk 3]
```diff
@@ -277,7 +290,7 @@ def build_through_table(
         if relation is None and table.is_temporal:
             relation = field.relation
         right_dataset_id, right_table_id = [
-            to_snake_case(part) for part in relation.split(":")[:2]
+            to_snake_case(part) for part in str(relation).split(":")[:2]
```
Comment (vdboor) [inline src/schematools/types.py:293]: Dit wordt dus `"None".split(":")[:2]` en geeft nog vreemdere situaties zo.

File: src/schematools/types.py [hunk 4]
```diff
@@ -24,11 +37,11 @@ def __missing__(self, key: str) -> NoReturn:
 
     @property
     def id(self) -> str:
-        return self["id"]
+        return str(self["id"])
 
     @property
     def type(self) -> str:
-        return self["type"]
+        return str(self["type"])
```
Comment (vdboor) [inline src/schematools/types.py:44]: misschien liever `cast(str, self["type"])`?

File: src/schematools/types.py [hunk 5]
```diff
@@ -214,11 +227,11 @@ def build_nested_table(
         # we need to add a shortname to the dynamically generated
         # schema definition.
         if field.has_shortname or table.has_shortname:
-            snakecased_fieldname = to_snake_case(field.name)
+            snakecased_fieldname: str = to_snake_case(field.name)
```
Comment (vdboor) [inline src/schematools/types.py:230]: is dat nodig? niet leiver `to_snake_case()` voorzien van een return type?

File: src/schematools/types.py [hunk 6]
```diff
@@ -406,16 +426,25 @@ def __init__(
...
             raise ValueError("Invalid JSON-schema contents of table")
 
+    @classmethod
+    def from_dict(cls, obj: Dict[str, Any]) -> DatasetTableSchema:
+        """Create tableschema from a dict.
+
+        When this factory function is used, the _parent_schema argument is missing
+        and will be set to an empty DatasetSchema.
+        """
+        return cls(DatasetSchema({"id": "parent", "type": "dataset"}), obj)
```
Comment (vdboor) [inline src/schematools/types.py:436]: Is dat handig? straks kunnen we debuggen waarom er zo'n vreemde "parent" dataset rondspookt.

File: src/schematools/types.py [hunk 7]
```diff
@@ -406,16 +426,25 @@ def __init__(
...
+    def name(self) -> str:
         return self.get("shortname", self.id)
 
     @property
     def has_shortname(self) -> bool:
         return self.get("shortname") is not None
 
     @property
-    def dataset(self) -> Optional[DatasetSchema]:
+    def dataset(self) -> DatasetSchema:
```
Comment (vdboor) [inline src/schematools/types.py:447]: moet dan ook weer anders

File: src/schematools/types.py [hunk 8]
```diff
@@ -660,24 +688,35 @@ def __init__(
...
 
+    @classmethod
+    def from_dict(cls, obj: Dict[str, Any]) -> DatasetFieldSchema:
+        """Create fieldschema from a dict.
+
+        When this factory function is used, the _id and _parent_table arguments are missing
+        and will be set to the "id" of the obj argument and an empty DatasetSchema.
+        """
+        return cls(
+            obj["id"], DatasetTableSchema(DatasetSchema({"id": "parent", "type": "dataset"})), obj
```
Comment (vdboor) [inline src/schematools/types.py:699]: is dat handig, dat er een fake tabel in gezet wordt? Geeft ook weer andere rare fouten

File: src/schematools/types.py (hunk 1 above)
Comment (gkoller) [reply src/schematools/types.py:120]: There's more to it. Let me look something up

Comment (gkoller) [reply src/schematools/types.py:120]: No, `version` is a required property according to the meta schema. Hence we should always be able to return a version. Meaning `-> str` is the correct return type

File: src/schematools/types.py (hunk 6 above)
Comment (jjmurre) [reply src/schematools/types.py:436]: De id en parent waren optioneel, maar zijn eigenlijk overal nodig. Probleem als ze optioneel zijn is ook dat je een shitload extra checks nodig hebt om mypy tevreden te stellen.
Overigens wordt deze factory function voor de DatasetTable maar op 1 plek in de DSO gebruikt in een testje. We kunnen ook deze factory function weghalen. Die fake dataset zit mij idd ook niet lekker.

File: src/schematools/types.py (hunk 3 above)
Comment (jjmurre) [reply src/schematools/types.py:293]: Goed punt, beter met een cast.

File: src/schematools/types.py (hunk 4 above)
Comment (gkoller) [reply src/schematools/types.py:44]: The conversions or casts are necessary as MyPy has no idea what the types of the values associated with the keys `id` and `type` are. But I'm wondering if we can do better?

Schemas have a set a predefined attributes and a set of user definable attributes. If we would have had only predefined attributes we could have used a `dataclass` (or `TypedDict`). That way MyPy would have know that both `id` and `type` are of type`str`, obviating the need for casts/conversions.

However both `dataclass`es and `TypedDict`s do not support extra keys (the user definable attributes in the schemas, eg column names of tables). Hence we can't use a `dataclass` or even a `TypedDict` as the base class for our schema types. I'm wondering if there is anything out there that would allow for extra keys in Python data structures while still providing the type checking for the predefined keys?

File: src/schematools/types.py [hunk 9]
```diff
@@ -37,7 +50,7 @@ def json_data(self) -> Dict[str, Any]:
         return self.data
 
     @classmethod
-    def from_dict(cls, obj: Dict[str, Any]) -> ST:
+    def from_dict(cls, obj: Dict[str, Any]) -> SchemaType:
```
Comment (gkoller) [inline src/schematools/types.py:53]: This is incorrect. `SchemaType.from_dict` is inherited by more specialized classes, eg `DatasetSchema`. If we call `DatasetSchema.from_dict` we do **not** want it to return a `SchemaType`, we want it to return a `DatasetSchema`. The `TypeVar` `ST` ensure exactly that.

Comment (gkoller) [reply src/schematools/types.py:53]: I'm sorry. Ignore my comment. `-> SchemaType` is correct.

File: src/schematools/types.py (hunk 1 above)
Comment (gkoller) [reply src/schematools/types.py:120]: No, version is a required property according to the meta schema. Hence we should always be able to return a version. Meaning -> str is the correct return type

File: src/schematools/types.py (hunk 5 above)
Comment (gkoller) [reply src/schematools/types.py:230]: `to_snake_case()` is already type annotated to return a `str` Hence MyPy should be able to derive this.

File: src/schematools/types.py [hunk 10]
```diff
@@ -286,7 +299,7 @@ def build_through_table(
         snakecased_field_id = to_snake_case(field.id)
         table_id = get_rel_table_identifier(len(self.id) + 1, table.id, snakecased_field_id)
 
-        sub_table_schema = {
+        sub_table_schema: Dict[str, Any] = {
```
Comment (gkoller) [inline src/schematools/types.py:302]: This is a recurring datatype. Maybe introduce an alias for it?

```python
Json = Dict[str, Any]
```

Or to provide some top level type checking until MyPy support recursive types:

```python
Json = Union[Dict[str, Any], List, str, int, float, bool, None, object]
```

Yes `Json` instead of `JSON`. Why? Loved this explanation in https://nim-lang.org/docs/nep1.html

> In the age of HTTP, HTML, FTP, TCP, IP, UTF, WWW it is foolish to pretend these are somewhat special words requiring all uppercase. Instead treat them as what they are: Real words. So it's parseUrl rather than parseURL, checkHttpHeader instead of checkHTTPHeader etc.

File: src/schematools/types.py [hunk 11]
```diff
@@ -475,12 +503,12 @@ def get_through_tables_by_id(self) -> List[DatasetTableSchema]:
         return tables
 
     @property
-    def display_field(self):
+    def display_field(self) -> Optional[Any]:
```
Comment (gkoller) [inline src/schematools/types.py:506]: Should this be `-> Optional[str]`

I am suggesting `str` as `display` is defined in the meta schema as:

```json
    "display": {
      "type": "string"
    },
```
However the docstring below suggest the result, if any, can be plural (fields). Should it be `Optional[List[str]]` then?

File: src/schematools/types.py [hunk 12]
```diff
@@ -475,12 +503,12 @@ def get_through_tables_by_id(self) -> List[DatasetTableSchema]:
...
         """Tell which fields can be used as display field."""
         return self["schema"].get("display", None)
 
-    def get_dataset_schema(self, dataset_id):
-        """Return another datasets"""
+    def get_dataset_schema(self, dataset_id: str) -> Optional[DatasetSchema]:
+        """Return another dataset"""
         return self._parent_schema.get_dataset_schema(dataset_id)
```
Comment (gkoller) [inline src/schematools/types.py:512]: How does this function work? Every `DatasetSchema` has a `_parent_schema` of type DatasetSchema that is required (it is not `Optional`)

Calling `get_dataset_schema` on a `DatasetSchema` instance recursively calls `get_dataset_schema` on its parent. How does it stop? In other words what is its stopping condition?

File: src/schematools/types.py [hunk 13]
```diff
@@ -536,43 +564,45 @@ def identifier(self) -> List[str]:
         # Convert identifier to a list, to be backwards compatible with older schemas
         if not isinstance(identifier, list):
             identifier = [identifier]
-        return identifier
+        return list(identifier)  # mypy pleaser
```
Comment (gkoller) [inline src/schematools/types.py:567]: Amazed this is needed to please MyPy. What version are you running?

And if it is really needed a `cast(List[str], ....)` would arguably be more appropriate.

File: src/schematools/types.py [hunk 14]
```diff
@@ -536,43 +564,45 @@ def identifier(self) -> List[str]:
...
     @property
-    def filters(self):
+    def filters(self) -> Dict[str, Dict[str, str]]:
         """Fetch list of additional filters"""
-        return self["schema"].get("additionalFilters", {})
+        return dict(self["schema"].get("additionalFilters", {}))
 
     @property
-    def relations(self):
+    def relations(self) -> Dict[Any, Any]:
```
Comment (gkoller) [inline src/schematools/types.py:595]: I am not so familiar with relations. However `Any` for the key seems especially broad. I assume it can be `str` if I read the docstring correctly (_dictionary of names ..._)

Regarding the values. Does it resolve relations to actual table definitions? If so the return type can be `Dict[str, DatasetTableSchema]`.

If on the other hand it simple returns a JSON blob, you could turn it into `Dict[str, Json]` using the `Json` definition in https://github.com/Amsterdam/schema-tools/pull/159/files#r653344045

File: src/schematools/types.py [hunk 15]
```diff
@@ -594,7 +624,7 @@ def model_name(self) -> str:
         model_name = self.id
         if self.dataset.version is not None and not self.dataset.is_default_version:
             model_name = f"{model_name}_{self.dataset.version}"
-        return to_snake_case(model_name)
+        return str(to_snake_case(model_name))  # mypy pleaser
```
Comment (gkoller) [inline src/schematools/types.py:627]: Should not be necessary. `to_snake_case` is already type annotated to return a `str`.

File: src/schematools/types.py [hunk 16]
```diff
@@ -754,18 +793,18 @@ def provenance(self) -> Optional[str]:
...
     def get_dimension_fieldnames_for_relation(
-        self, relation: str, nm_relation: str
+        self, relation: Optional[str], nm_relation: Optional[str]
     ) -> Dict[str, List[str]]:
         """Gets the dimension fieldnames."""
         if relation is None and nm_relation is None:
             return {}
 
-        dataset_id, table_id = (relation or nm_relation).split(":")
+        dataset_id, table_id = str(relation or nm_relation).split(":")
```
Comment (gkoller) [inline src/schematools/types.py:807]: Hmm, so I presume MyPy can't determine, from the above if statement, that this or-expression evaluates to a `str`? Weird.

File: src/schematools/types.py [hunk 17]
```diff
@@ -892,25 +931,25 @@ class ProfileSchema(SchemaType):
...
+    def name(self) -> Optional[str]:
         """Name of Profile (if set)"""
         return self.get("name")
 
     @property
-    def scopes(self):
+    def scopes(self) -> List[str]:
         """Scopes of Profile (if set)"""
-        return self.get("scopes")
+        return list(self.get("scopes", []))
```
Comment (gkoller) [inline src/schematools/types.py:952]: `cast(List[str], ...)`

File: src/schematools/utils.py [hunk 18]
```diff
@@ -59,15 +67,22 @@ def profile_defs_from_url(profiles_url: Union[URL, str]) -> Dict[str, types.Prof
...
+        for p_name, p_schema in defs_from_url(
+            base_url=profiles_url, data_type=types.ProfileSchema
+        ).items()
+    }
 
 
-def defs_from_url(base_url: Union[URL, str], data_type: Type[types.ST]) -> Dict[str, types.ST]:
+def defs_from_url(
+    base_url: Union[URL, str], data_type: Type[types.SchemaType]
+) -> Dict[str, types.SchemaType]:
```
Comment (gkoller) [inline src/schematools/utils.py:80]: You should revert this change. By returning `SchemaType` here you need all those explicit upcasts in `schema_def_from_url` and `schema_defs_from_url`. With the original `TypeVar` definition `ST` those upcasts (from `SchemaType` to more specific type`) will not be necessary anymore.

File: src/schematools/utils.py [hunk 19]
```diff
@@ -59,15 +67,22 @@ def profile_defs_from_url(profiles_url: Union[URL, str]) -> Dict[str, types.Prof
...
     """Fetch all schema definitions from a remote file.
 
     The URL could be ``https://schemas.data.amsterdam.nl/datasets/``
     """
-    schema_lookup: Dict[str, types.ST] = {}
+    schema_lookup: Dict[str, types.SchemaType] = {}
```
[the remark this answers is not in the document]
Comment (jjmurre) [reply src/schematools/utils.py:85]: Hm. Whenever used `ST` mypy was always complaining about it. Could this be a mypy version problem?

File: src/schematools/utils.py [hunk 20]
```diff
@@ -109,7 +124,7 @@ def def_from_url(
         response = connection.get(base_url / index[dataset_id])
         response.raise_for_status()
 
-        dataset_schema: types.ST = data_type.from_dict(response.json())
+        dataset_schema: types.SchemaType = data_type.from_dict(response.json())
```
Comment (gkoller) [inline src/schematools/utils.py:127]: `ST` was correct. By downcasting every possible `SchemaType` to `SchemaType` here, you'll need upcasts to more specific types elsewhere. `ST`'s definition prevents that.

File: src/schematools/types.py (hunk 12 above)
Comment (jjmurre) [reply src/schematools/types.py:512]: That is not how it works. The `DatasetTableSchema` has a `_parent_schema` which is a `DatasetSchema`. So, there is no recursion involved.

File: src/schematools/types.py (hunk 13 above)
Comment (jjmurre) [reply src/schematools/types.py:567]: My mypy version is 0.812 (feb 19, 2021), seems quite recent. I agree, cast is better.

File: src/schematools/types.py (hunk 4 above)
Comment (gkoller) [reply src/schematools/types.py:46]: Pydantic seems to be able to provide the means to do exactly that with the following option:
```python
    class Config:
        extra = 'allow'
```

See also: https://pydantic-docs.helpmanual.io/usage/model_config/

* extra
  *  whether to ignore, allow, or forbid extra attributes during model initialization. Accepts the string values of 'ignore', 'allow', or 'forbid', or values of the Extra enum (default: Extra.ignore). 'forbid' will cause validation to fail if extra attributes are included, 'ignore' will silently ignore any extra attributes, and 'allow' will assign the attributes to the model.

File: src/schematools/types.py (hunk 15 above)
Comment (jjmurre) [reply src/schematools/types.py:629]: Yes:
    returning Any from function declared to return "str"  [no-any-return]
Now solved it with a `cast`. I think you are right (other comment) that the return type of `to_snake_case`, because of the decorator, cannot be inferred correctly.

File: src/schematools/types.py (hunk 10 above)
Comment (vdboor) [reply src/schematools/types.py:304]: Grappig, ik die ook altijd `checkHttpUrl` i.p.v. `checkHTTPURL`, maar als er niet geconcat wordt wel `URL` en dus `JSON` :-p. Hoewel ik wel weer een `QUrl` gebruik.

File: src/schematools/utils.py [hunk 21]
```diff
@@ -59,7 +64,12 @@ def profile_defs_from_url(profiles_url: Union[URL, str]) -> Dict[str, types.Prof
 
     The URL could be ``https://schemas.data.amsterdam.nl/profiles/``
     """
-    return defs_from_url(base_url=profiles_url, data_type=types.ProfileSchema)
+    return {
+        p_name: p_schema
+        for p_name, p_schema in defs_from_url(
+            base_url=profiles_url, data_type=types.ProfileSchema
+        ).items()
+    }
```
Comment (vdboor) [inline src/schematools/utils.py:72]: Hebben we dit echt nodig?

File: src/schematools/utils.py [hunk 22]
```diff
@@ -36,7 +36,12 @@ def schema_defs_from_url(
         )
         return {dataset_name: schema}
 
-    return defs_from_url(base_url=schemas_url, data_type=types.DatasetSchema)
+    return {
+        ds_name: ds_schema
+        for ds_name, ds_schema in defs_from_url(
+            base_url=schemas_url, data_type=types.DatasetSchema
+        ).items()
+    }
```
Comment (vdboor) [inline src/schematools/utils.py:44]: Hebben we dit echt nodig?

File: src/schematools/types.py [hunk 23]
```diff
@@ -952,7 +998,12 @@ def tables(self) -> Dict[str, ProfileTableSchema]:
 class ProfileTableSchema(DatasetType):
     """A single table in the profile"""
 
-    def __init__(self, _id, _parent_schema=None, data: Optional[Dict[str, Any]] = None):
+    def __init__(
+        self,
+        _id: str,
+        _parent_schema: Optional[ProfileDatasetSchema] = None,
+        data: Optional[Dict[str, Any]] = None,
```
Comment (vdboor) [inline src/schematools/types.py:1005]: data niet gewoon verplicht maken i.p.v. optional?

File: src/schematools/types.py [hunk 24]
```diff
@@ -923,7 +964,12 @@ def datasets(self) -> Dict[str, ProfileDatasetSchema]:
 class ProfileDatasetSchema(DatasetType):
     """A schema inside the profile dataset"""
 
-    def __init__(self, _id, _parent_schema=None, data: Optional[Dict[str, Any]] = None):
+    def __init__(
+        self,
+        _id: str,
+        _parent_schema: Optional[ProfileSchema] = None,
+        data: Optional[Dict[str, Any]] = None,
```
Comment (vdboor) [inline src/schematools/types.py:971]: Kunnen we data verplicht maken?

File: src/schematools/types.py (hunk 12 above)
Comment (gkoller) [reply src/schematools/types.py:514]: Ah, thanks for the clarification. I had missed that this was `DatasetTableSchema`'s `get_dataset_schema`.

File: src/schematools/types.py (hunk 6 above)
Comment (jjmurre) [reply src/schematools/types.py:438]: This fake parent was not a good idea. I reverted it and made the parameters optional again.

File: src/schematools/types.py (hunk 7 above)
Comment (jjmurre) [reply src/schematools/types.py:449]: Restored to Optional. And added extra checks for None in the code where `dataset` is used.

File: src/schematools/types.py (hunk 8 above)
Comment (jjmurre) [reply src/schematools/types.py:701]: Verwijderd.

File: src/schematools/utils.py (hunk 21 above)
Comment (jjmurre) [reply src/schematools/utils.py:72]: Kon gelukkig weer verwijderd worden agv. fix door @gkoller bij het gebruik van `ST` typevar.

File: src/schematools/utils.py (hunk 22 above)
Comment (jjmurre) [reply src/schematools/utils.py:44]: Niet meer, zie hierboven.

File: src/schematools/types.py (hunk 23 above)
Comment (jjmurre) [reply src/schematools/types.py:1005]: Zo te zien wel. Test blijven werken, ook DSO API tests blijven werken. Aangepast.

File: src/schematools/utils.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (jjmurre) [reply src/schematools/utils.py]: You fixed it with an extra commit on the PR, tx!

File: src/schematools/utils.py (hunk 18 above)
Comment (jjmurre) [reply src/schematools/utils.py]: Fixed!

File: src/schematools/types.py (hunk 16 above)
Comment (jjmurre) [reply src/schematools/types.py:809]: Using cast now.

File: src/schematools/types.py (hunk 14 above)
Comment (jjmurre) [reply src/schematools/types.py:597]: Has always a fixed structure, changed it to `Dict[str, Dict[str, str]]`.

File: src/schematools/types.py (hunk 10 above)
Comment (jjmurre) [reply src/schematools/types.py:304]: I tried the approach with the `Json` alias. The first (simple) variation works. However, for the `Union`  type it is not just introducing the alias and doing a search-replace. This generates 19 mypy new errors. So this Union needs some more investigation.

File: src/schematools/types.py (hunk 11 above)
Comment (jjmurre) [reply src/schematools/types.py:508]: It should indeed be `Optional[str]`. Changed.

File: src/schematools/types.py (hunk 3 above)
Comment (jjmurre) [reply src/schematools/types.py:295]: Added extra precaution, relation should never be None.

File: src/schematools/types.py (hunk 14 above)
Comment (gkoller) [reply src/schematools/types.py]: You might consider making it an alias:
```python
Relations = Dict[str, Dict[str, str]]
...
def relations(self) -> Relations:
    ....
```
This way a future more strict (dataclass/Pydantic) based type can easily be introduced by search and replacing `Relations` with `<new-fancy-name-for-new-type>` and addressing the issues MyPy then finds.

**Review by gkoller — APPROVED** [review]
Excellent improvement! :+1:
